paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+        output = []
+        # Wait until process terminates...
+        while p.poll() is None:
+            line = p.communicate()[0]
+            output.append(line)
+        if p.poll() is None:
+            p.kill()
+        self.grep_leap_log()
+        return output
+
+    def grep_leap_log(self):
+        """
+        Check for a few keywords in the `tleap` output.
+
+        """
+        try:
+            with open(self.output_path + 'leap.log', 'r') as file:
+                for line in file.readlines():
+                    if re.search('ERROR|WARNING|Warning|duplicate|FATAL|Could', line):
+                        log.warning(
+                            'It appears there was a problem with solvation: check `leap.log`...'
+                        )
+        except:
+            return
+
+    def check_for_leap_log(self):
+        """
+        Check if `leap.log` exists in output_path, and if so, delete so the current run doesn't append.
+
+        """
+        filename = 'leap.log'
+        try:
+            os.remove(self.output_path + filename)
+            log.debug('Deleted existing leap.log file: '+self.output_path + filename)
+        except OSError:
+            pass
+
+    def solvate(self):
+        """
+        Solvate a structure with an exact number of waters or buffer size.
+
+        """
+    
+        # If `buffer_target` is a string ending with 'A', an estimate of the number of waters is generated, otherwise,
+        # the target is returned.
+        self.set_target_waters()
+    
+        if self.add_ions:
+            self.set_additional_ions()
+    
+        # First, a coarse adjustment...
+        # This will run for 50 iterations or until we (a) have more waters than the target and (b) are within ~12 waters
+        # of the target (that can be manually removed).
+        cycle = 0
+        while cycle < self.max_cycles:
+            # Start by not manually removing any water, just adjusting the buffer value to get near the target number of
+            # waters...
+
+            # Find out how many waters for *this* buffer value...
+            waters = self.count_waters()
+            self.wat_added_history.append(waters)
+            self.buffer_val_history.append(self.buffer_value)
+            log.debug("Cycle {:02.0f}   {:8.5f} {:6.0f} ({:6.0f})".format(cycle, self.buffer_value, waters, self.target_waters))
+    
+            # Possible location of a switch to adjust the buffer_value by polynomial
+            # fit approach.
+    
+            # If we've nailed it, break!
+            if waters == self.target_waters:
+                # Run one more time and save files
+                self.write_input(include_saves=True)
+                self.run()
+                return
+            # If we are close, go to fine adjustment...
+            elif waters > self.target_waters and (waters - self.target_waters) < 12:
```

Comment:
I've been wondering if the 12 manually deleted waters could be configurable. As this number goes up, the density gets away from 1.00 g/liter but for a huge system, you could save time by making this 20 or 30 waters, right?

Reply:
Yep.  This could be an advanced class variable.

Reply:
Done.  Created `self.max_cycles`